### PR TITLE
Add address_of macro

### PR DESCRIPTION
**Pull request by teggotic**
Fixes https://github.com/jemc/mare/issues/14

### CODE REVIEW

File: src/mare/compiler/macros.cr [hunk 1]
```diff
@@ -139,6 +139,12 @@ class Mare::Compiler::Macros < Mare::AST::CopyOnMutateVisitor
         "the value whose identity is to be hashed",
       ])
       visit_identity_digest_of(node)
+    elsif Util.match_ident?(node, 0, "address_of")
+      Util.require_terms(node, [
+        nil,
+        "the value whose address is to be gotten",
```
Comment (jemc) [inline src/mare/compiler/macros.cr:145]:
```suggestion
        "the local variable whose address is to be referenced by an FFI call",
```

File: src/mare/compiler/refer.cr [hunk 2]
```diff
@@ -102,6 +102,15 @@ module Mare::Compiler::Refer
       @analysis[node] = info
     end
 
+    def touch(ctx, node : AST::Prefix)
+      case node.op.value
+      when "address_of"
+        unless node.term.is_a? AST::Identifier
```
Comment (jemc) [inline src/mare/compiler/refer.cr:108]: I'm thinking that we need a different check for this:
```suggestion
        unless @analysis[node.term].is_a? Local
```

An identifier could be many different kinds of things, such as a type name for example.

We want to check there that it is specifically a local variable.

File: src/mare/compiler/refer.cr [hunk 3]
```diff
@@ -102,6 +102,15 @@ module Mare::Compiler::Refer
...
+          Error.at node.term, "address_of can be applied only to variable"
```
Comment (jemc) [inline src/mare/compiler/refer.cr:109]: Can you add a test to `refer_spec.cr` that demonstrates this error?

If we don't have tests for it I could accidentally break it at any time and not know that I did.

File: src/mare/compiler/macros.cr [hunk 4]
```diff
@@ -139,6 +139,12 @@ class Mare::Compiler::Macros < Mare::AST::CopyOnMutateVisitor
         "the value whose identity is to be hashed",
       ])
       visit_identity_digest_of(node)
+    elsif Util.match_ident?(node, 0, "address_of")
```
Comment (jemc) [inline src/mare/compiler/macros.cr:142]: Can you add a `spec/compiler/macros/address_of_spec.cr` file that has tests for this?

You should be able to copy tests from `spec/compiler/macros/identity_digest_of_spec.cr` and just adjust the text.

File: src/mare/compiler/code_gen.cr [hunk 5]
```diff
@@ -1893,6 +1893,15 @@ class Mare::Compiler::CodeGen
     end
   end
 
+  def gen_address_of(term_expr)
+    # term_expr = term_expr.as AST::Identifier
```
Comment (jemc) [inline src/mare/compiler/code_gen.cr:1897]: Looks like this line is no longer needed.
```suggestion
```

File: src/mare/compiler/infer/info.cr [hunk 6]
```diff
@@ -454,6 +454,24 @@ class Mare::Compiler::Infer
     end
   end
 
+  class AddressOf < DynamicInfo
+    getter variable : Info
+
+    def describe_kind; "address of" end
```
Comment (jemc) [inline src/mare/compiler/infer/info.cr:460]: This description will lead to slightly nicer typecheck error messages:
```suggestion
    def describe_kind; "address of this variable" end
```

File: src/mare/compiler/macros.cr (hunk 1 above)
Comment (teggotic) [reply src/mare/compiler/macros.cr:145]: Actually we don't have a restriction for ffi only.

File: src/mare/compiler/refer.cr (hunk 2 above)
Comment (teggotic) [reply src/mare/compiler/refer.cr:108]: Sure

File: src/mare/compiler/macros.cr (hunk 1 above)
Comment (jemc) [reply src/mare/compiler/macros.cr:145]: I think we probably should, eventually. But point taken - I will remove that part of the suggestion for now.

Comment (teggotic) [reply src/mare/compiler/macros.cr:145]: What about having something like unsafe block? We can have something like unsafe it rust, the function can be unsafe and it means that we can use it in unsafe block or function, and the unsafe block to use unsafe functions or address_of feature. I mean that's strange for me that we allow unsafe `"".cpointer()` but we don't allow unsafe `foo = address_of bar`.

File: src/mare/compiler/refer.cr (hunk 3 above)
Comment (teggotic) [reply src/mare/compiler/refer.cr:109]: Sure

File: src/mare/compiler/macros.cr (hunk 4 above)
Comment (teggotic) [reply src/mare/compiler/macros.cr:142]: Sure

**Comment by jemc** [comment]
Thanks!

**Pull request merged**